File: Practice/src/hash/UnfinishedPlayer.js
function solution(participant, completion) {
    const map = new Map();
    for (let i of participant) {
        map.set(i, map.has(i) ? map.get(i) + 1 : 1);
    }
    for (let i of completion) {
        map.set(i, map.get(i) - 1);
    }
    // for (let i = 0; i < participant.length; i++) {
    //     let a = participant[i],
    //         b = completion[i];
    //     map.set(a, (map.get(a) || 0) + 1);
    //     map.set(b, (map.get(b) || 0) - 1);
    // }
    for (let [k, v] of map) {
        if (v) return k;
    }
    // participant.sort();
    // completion.sort();
    // for (var i = 0; i < participant.length; i++) {
    //     if (participant[i] != completion[i]) {
    //         return participant[i];
    //     }
    // }
}
// var solution = (participant, completion) =>
//     participant.find(
//         (participant) => !completion[participant]--,
//         completion.map(
//             (participant) =>
//                 (completion[participant] = (completion[participant] | 0) + 1)
//         )
//     );

File: Practice/src/hash/UnfinishedPlayer.java
package hash;

import java.util.HashMap;
import java.util.Map.Entry;

public class UnfinishedPlayer {
//	완주하지 못한 선수
	public String solution(String[] participant, String[] completion) {
		HashMap<String, Integer> marathon = new HashMap<String, Integer>();
		for (String s : participant) {
			marathon.put(s, marathon.getOrDefault(s, 0) + 1);
		}
		for (String s : completion) {
			marathon.put(s, marathon.get(s) - 1);
		}
		for (Entry<String, Integer> ent : marathon.entrySet()) {
			if (ent.getValue() == 1) {
				return ent.getKey();
			}
		}
		return null;
//		for (int i = 0; i < participant.length; i++) {
//			marathon.put(participant[i], getHashValue(marathon, participant[i]) + 1);
//		}
//		for (int i = 0; i < completion.length; i++) {
//			marathon.put(completion[i], getHashValue(marathon, completion[i]) - 1);
//		}
//		// marathon.forEach((k,v) -> { if(v==1) return ""; });
//		for (Map.Entry<String, Integer> entry : marathon.entrySet()) {
//			if (entry.getValue() == 1)
//				return entry.getKey();
//		}
//		return "";
	}

//	public static int getHashValue(HashMap<String, Integer> hash, String check) {
//		return hash.get(check) == null ? 0 : hash.get(check);
//	}
}